src/Interface/hiopInterfacePrimalDecomp.cpp

```diff
@@ -0,0 +1,216 @@
+#include "hiopInterfacePrimalDecomp.hpp"
+
+
+using namespace hiop;
+hiopInterfacePriDecProblem::RecourseApproxEvaluator::
+RecourseApproxEvaluator(int nc) : RecourseApproxEvaluator(nc, nc)  //nc_ <= nx, nd=S
+{
+}
+
+hiopInterfacePriDecProblem::RecourseApproxEvaluator::
+RecourseApproxEvaluator(int nc, int S) 
+  : nc_(nc),S_(S),rval_(0.)//nc = nx, nd=S
+{
+  assert(S>=nc);
+  xc_idx_.resize(nc_);
+  for(int i=0;i<nc_;i++) xc_idx_[i] = i;
+  rgrad_ = new double[nc];
+  rhess_ = new double[nc];
+  x0_ = new double[nc];
+}
+
+hiopInterfacePriDecProblem::RecourseApproxEvaluator::
+RecourseApproxEvaluator(const int nc, 
+                        const int S, 
+                        const std::vector<int>& list)
+  : nc_(nc),S_(S),rval_(0.),rgrad_(NULL), rhess_(NULL),x0_(NULL)//nc = nx, nd=S
+{
+  rgrad_ = new double[nc];
+  rhess_ = new double[nc];
+  x0_ = new double[nc];
+  assert(list.size()==nc_);
+  xc_idx_ = list;
+}
+
+hiopInterfacePriDecProblem::RecourseApproxEvaluator::
+RecourseApproxEvaluator(const int nc, 
+                        const int S, 
+                        const double& rval, 
+                        const double* rgrad, 
+                        const double* rhess, 
+                        const double* x0)
+  : nc_(nc), S_(S)
+{
+  //assert(S>=nc);
+  rval_ = rval;
+  rgrad_ = new double[nc];
+  rhess_ = new double[nc];
+  x0_ = new double[nc];
+    
+  xc_idx_.resize(nc_);
+  for(int i=0;i<nc_;i++) xc_idx_[i] = i;
+
+  memcpy(rgrad_,rgrad, nc*sizeof(double));
+  memcpy(rhess_,rhess , nc*sizeof(double));
+  memcpy(x0_,x0, nc*sizeof(double));
+}
+
+hiopInterfacePriDecProblem::RecourseApproxEvaluator::
+RecourseApproxEvaluator(const int nc, 
+                        const int S, 
+                        const std::vector<int>& list,
+                        const double& rval, 
+                        const double* rgrad, 
+                        const double* rhess, 
+                        const double* x0)
+  : nc_(nc), S_(S)
+{
+  //assert(S>=nc);
+  rval_ = rval;
+  rgrad_ = new double[nc];
+  rhess_ = new double[nc];
+  x0_ = new double[nc];
+  assert(list.size()==nc_);
+  xc_idx_ = list;
+
+  memcpy(rgrad_,rgrad, nc*sizeof(double));
+  memcpy(rhess_,rhess , nc*sizeof(double));
+  memcpy(x0_,x0, nc*sizeof(double));
+}
+
+/**
+ * x is the full optimization variable for the base problem
+ * x0 only stores the coupled x
+ * Therefore need to pick out the coupled ones
+ * n is the total dimension of x and not really used in the function
+ */
+bool hiopInterfacePriDecProblem::RecourseApproxEvaluator::
+eval_f(const long long& n, const double* x, bool new_x, double& obj_value)
+{
+  //assert(nc>=4);
+  //sum 0.5 {(x_i-1)*(x_{i}-1) : i=1,...,nc} 
+  assert(rgrad_!=NULL);
+  obj_value += rval_;
+
+  for(int i=0; i<nc_; i++) {
+    obj_value += rgrad_[i]*(x[xc_idx_[i]]-x0_[i]);
+  }
+  for(int i=0; i<nc_; i++) {
+    obj_value += 0.5*rhess_[i]*(x[xc_idx_[i]]-x0_[i])*(x[xc_idx_[i]]-x0_[i]);
+  }
```

Comment:
The code in this implementation is not portable. I will not mark each instance, but rather stop here. It should be fairly straightforward to reimplement this in a portable fashion using existing linear algebra objects in HiOp.

Reply:
OK. Will do.

Reply:
Just to make sure I understand, should I for example use LinearAlgebraFactory::createVector(n) for all the double * ? And all the protected/private variables should be hiopVector instead of double*? Thanks.

Reply:
I think the most important part is separating platform specific code from the algorithm implementation. Platform specific code should be accessed through an abstract interface.

I think encapsulating data arrays in `hiopVector` objects is a good idea and vectors should be created using factory. We may need to update the `LinearAlgebraFactory` to meet requirements of the primal decomposition module.

Reply:
Sounds good. Doing the conversion to hiopVector now.

Reply:
The only part not converted yet is the x[xc_idx_[i]] part, which access x through a list of integer vector. We should discuss this further.

Reply:
@jwang125, you are right. We don't have a good portable example of re-mapping vectors in HiOp code.